src/language/mongoDBService.ts

```diff
@@ -203,200 +212,98 @@ export default class MongoDBService {
+   */
+  async _getCollections(databaseName: string): Promise<Document[]> {
+    let result: Document[] = [];
 
-          void worker.terminate().then(() => {
-            this._connection.console.log(
-              `MONGOSH found ${result.length} collections`
-            );
-            this._updateCurrentSessionCollections(databaseName, result);
+    if (!this._serviceProvider) {
+      return [];
+    }
 
-            return resolve(true);
-          });
-        });
-      } catch (error) {
-        this._connection.console.log(
-          `MONGOSH get list collections error: ${util.inspect(error)}`
-        );
+    try {
+      const documents = await this._serviceProvider.listCollections(
+        databaseName
+      );
+      result = documents ? documents : [];
+    } catch (error) {
+      this._connection.console.error(`LS get collections error: ${error}`);
+    }
 
-        return resolve(false);
-      }
-    });
+    return result;
   }
 
-  _getFieldsCompletionItems(
+  /**
+   * Get field names for the provided namespace.
+   */
+  async _getSchemaFields(
     databaseName: string,
     collectionName: string
-  ): Promise<boolean> {
-    return new Promise((resolve) => {
-      if (!this._extensionPath) {
-        this._connection.console.log(
-          'SCHEMA error: extensionPath is undefined'
-        );
-
-        return resolve(false);
-      }
-
-      try {
-        const namespace = `${databaseName}.${collectionName}`;
-        const worker = new WorkerThreads(
-          path.resolve(
-            this._extensionPath,
-            'dist',
-            languageServerWorkerFileName
-          ),
-          {
-            workerData: {
-              connectionString: this._connectionString,
-              connectionOptions: this._connectionOptions,
-              databaseName,
-              collectionName,
-            },
-          }
-        );
-
-        this._connection.console.log(`SCHEMA for namespace: "${namespace}"`);
-        worker.postMessage(ServerCommands.GET_FIELDS_FROM_SCHEMA);
+  ): Promise<string[]> {
+    let result: string[] = [];
 
-        worker.on('message', (response: [Error, CompletionItem[] | []]) => {
-          const [error, result] = response;
-
-          if (error) {
-            this._connection.console.log(
-              `SCHEMA error: ${util.inspect(error)}`
-            );
-          }
-
-          void worker.terminate().then(() => {
-            this._connection.console.log(
-              `SCHEMA found ${result.length} fields`
-            );
-            this._updateCurrentSessionFields(namespace, result);
+    if (!this._serviceProvider) {
+      return [];
+    }
 
-            return resolve(true);
-          });
-        });
-      } catch (error) {
-        this._connection.console.log(`SCHEMA error: ${util.inspect(error)}`);
+    try {
+      const documents = await this._serviceProvider
+        .find(databaseName, collectionName, {}, { limit: 1 })
```

Comment:
Unrelated note, should this really be `1`? `parseSchema()` is generally meant for input with more than one document, since otherwise it just gives you the (potentially nested) keys of the document (and it actually accepts a cursor, which means that the `.toArray()` call here might be unnecessary).

Reply:
As I remember we wanted to make the parsing process faster by parsing one document only. Maybe I miss something, but isn't Compass does the same? It uses just one document for this:
[compass
](https://github.com/mongodb-js/compass/blob/main/packages/compass-field-store/src/stores/store.js#L193)

And then passes it to the `parseSchema`:
https://github.com/mongodb-js/compass/blob/main/packages/compass-field-store/src/stores/store.js#L180

About `toArray` you are right, we can remove it.

Reply:
Interesting, I’m also surprised that Compass does this then (it even has multiple documents available in the refresh case… it makes sense that when inserting a document, it only processes that document, but otherwise I’m not sure why we use `parseSchema` then… :slightly_smiling_face:)

Reply:
I think we use parseSchema with one document to guess how the whole collection can look like and not parse a big amount of data, and also get types and other info that parseSchema results. Is there another way to get a field type?

Reply:
> and also get types and other info that parseSchema results

Yeah, that might change the equation a bit – I’m just surprised because below we do `schema.fields.map((item) => item.name)` to get the result of this function, so it seems like we currently only use the names?

To be clear, I don’t mind using `parseSchema`, I’m mostly just surprised that we’re only passing a single document here

Reply:
I see your point! Do you think it is worth to get rid of parseSchema and just get deep object keys?

Reply:
I’m not sure? Just to reiterate, it’s surprising to me that we don’t use more documents (not a lot – but maybe 5 or 10 or so?) than 1. And if we want to keep the flexibility to do so, then keeping `parseSchema` seems like the right thing here

Reply:
Left TODO comment in code and added the same question to the `Technical Alignment Sync` backlog)